sklearn/meson.build

```diff
@@ -0,0 +1,193 @@
+
+- If your compiler supports OpenMP but you still see this
+  message, please submit a bug report at:
+
+    https://github.com/scikit-learn/scikit-learn/issues
+
+- The build will continue with OpenMP-based parallelism
+  disabled. Note however that some estimators will run in
+  sequential mode instead of leveraging thread-based
+  parallelism.
+
+                    ***
+''')
+endif
+
+cc = meson.get_compiler('c')
+
+# Deal with M_PI & friends; add `use_math_defines` to c_args
+# Cython doesn't always get this correctly itself
+# explicitly add the define as a compiler flag for Cython-generated code.
+if is_windows
+  use_math_defines = ['-D_USE_MATH_DEFINES']
+else
+  use_math_defines = []
+endif
+
+# Don't use the deprecated NumPy C API. Define this to a fixed version instead of
+# NPY_API_VERSION in order not to break compilation for released SciPy versions
+# when NumPy introduces a new deprecation. Use in a meson.build file::
+#
+#   py3.extension_module('_name',
+#     'source_fname',
+#     numpy_nodepr_api)
+#
+numpy_nodepr_api = '-DNPY_NO_DEPRECATED_API=NPY_1_9_API_VERSION'
+
+# python_sources = [
+#   '__init__.py',
+#   '__init__.pyi',
+#   'conftest.py',
+# ]
+
+# py.install_sources(
+#   python_sources,
+#   pure: false,
+#   subdir: 'skimage'
+# )
+
+# TODO what is the sklearn equivalent
+# cython_cli = find_program('_build_utils/cythoner.py')
+
+# cython_gen = generator(cython_cli,
+#   arguments : ['@INPUT@', '@OUTPUT@'],
+#   output : '@BASENAME@.c')
+
+# cython_gen_cpp = generator(cython_cli,
+#   arguments : ['@INPUT@', '@OUTPUT@', '--cplus'],
+#   output : '@BASENAME@.cpp')
+
+c_undefined_ok = ['-Wno-maybe-uninitialized']
+
+# Suppress warning for deprecated Numpy API.
+# (Suppress warning messages emitted by #warning directives).
+# Replace with numpy_nodepr_api after Cython 3.0 is out
+# TODO What to do about this?
+_cpp_Wno_cpp = cpp.get_supported_arguments('-Wno-cpp')
+cython_c_args += [_cpp_Wno_cpp, use_math_defines]
+cython_cpp_args = cython_c_args
+
+# TODO I should reuse cython_c_args and merge my own tweaks
+
+extensions = ['_isotonic']
+
+py.extension_module(
+  '_isotonic',
+  '_isotonic.pyx',
+  # TODO what about additional arguments like below
+  # install: false,
+  #   c_args: [
+  #     '-DNPY_NO_DEPRECATED_API=0',  # Cython still uses old NumPy C API
+  #     # Require 1.25+ to test datetime additions
+  #     '-DNPY_TARGET_VERSION=NPY_2_0_API_VERSION',
+  #   ],
+  cython_args: cython_args,
+  install: true,
+  subdir: 'sklearn',
+)
+
+# TODO Needed for Cython?
+__init__py = fs.copyfile('__init__.py')
+
+sklearn_dir = py.get_install_dir() / 'sklearn'
+
+subdir('__check_build')
+subdir('_loss')
+# utils needs to be quite early since its .pxd are used by plenty of other modules Cython
+subdir('utils')
+# TODO metrics needs to be to be before cluster since cluster Cython imports
+# depends on metrics Cython. Better way to do this in meson?
```

Comment:
this seems to answer the question I think: https://stackoverflow.com/questions/55107396/meson-targets-that-depend-on-subdir-siblings